File: computational_physics/Homework1/hw_1_1.jl
using Printf
#a

# Function for nth order polynomial interpolation
function nth_order_interpolation(x::Float64, x_data::Array, y_data::Array)::Float64
    n = length(x_data)-1
    interpolated_value = 0.0
    # f(x)=\sum_{k}^{n+1}f_k*P_k^{n}(x) 
    for i in 1:n+1
        term = y_data[i]
        for j in 1:n+1
            if i != j
                term *= (x - x_data[j]) / (x_data[i] - x_data[j])
            end
        end
        interpolated_value += term
    end
    
    return interpolated_value
end
function error_for_nth_polynomial(x::Float64, x_data::Array, y_data::Array)::Float64
    n = length(x_data)-1
    error=0.0
    #\delta f(x)=f^{n+1}(x)*(x-x1)(x-x2)..............(x-x_{n+1})/(n+1)!
    term=nth_order_interpolation(x,x_data,y_data)/factorial(n+1)
    for i in 1:n+1
        term*=(x-x_data[i])
    end
    error+=term
    return error
end

x_datas=[0.2, 0.3, 0.4, 0.5, 0.6]
y_datas = [0.0015681, 0.0077382, 0.023579, 0.054849, 0.10696]

@printf(" polynomial value = %16.12f error :%16.12f  \n", nth_order_interpolation(0.46,x_datas,y_datas),error_for_nth_polynomial(0.46,x_datas,y_datas))
@printf(" polynomial value = %16.12f error :%16.12f  \n", nth_order_interpolation(0.16,x_datas,y_datas),error_for_nth_polynomial(0.16,x_datas,y_datas))

#b 

function neville_lagrange_interpolation(x::Float64, xdata::Array, ydata::Array)
    n = length(xdata)-1
    function_value= zeros(Float64, n+1, n+1)
    function_value[:, 1] .= ydata
    #neville's interpolation P(x) = \frac{(x - x_j) P_{0,1,\cdots,j-1,j+1,\cdots,k}(x) - (x - x_i) P_{0,1,\cdots,i-1,i+1,\cdots,k}(x)}{(x_i - x_j)}
    for i in 2:n+1
        for j in i:n+1
            function_value[j, i] = ((x - xdata[j - i + 1]) * function_value[j, i - 1] - (x - xdata[j]) * function_value[j - 1, i - 1]) / (xdata[j] - xdata[j - i + 1])
        end
    end
    
    approximated_value = function_value[n+1, n+1]
    
    return Dict("Approximated value" => approximated_value, "Neville iterations table" => function_value)
end

x1=0.16
x2=0.46
result = neville_lagrange_interpolation(x1, x_datas,y_datas)
println("Approximated value at x = $x1: ", result["Approximated value"])
println("Neville iterations table:")
display(result["Neville iterations table"])
result_2= neville_lagrange_interpolation(x2,x_datas,y_datas)
println("Approximated value at x = $x2: ", result_2["Approximated value"])
println("Neville iterations table:")
display(result_2["Neville iterations table"])

function error_neville_lagrange_interpolation(x::Float64, x_data::Array, y_data::Array)::Float64
    result=neville_lagrange_interpolation(x,x_data,y_data)
    n=length(x_data)-1
    numerical_uncertainty=0.5*(abs(result["Neville iterations table"][n+1,n+1]-result["Neville iterations table"][n,n])
                            +abs(result["Neville iterations table"][n+1,n+1]-result["Neville iterations table"][n+1,n]))
    return numerical_uncertainty
end
@printf(" polynomial value =%16.12f in neville's method at x  = %4.2f numerical uncertainty :%16.12f  \n", result["Approximated value"],x1,error_neville_lagrange_interpolation(x1,x_datas,y_datas))
@printf(" polynomial value =%16.12f in neville's method at x  = %4.2f numerical uncertainty :%16.12f  \n", result_2["Approximated value"],x2,error_neville_lagrange_interpolation(x2,x_datas,y_datas))

#c

function linear_interpolation(x_values::Array, y_values::Array, x::Float64)::Float64
    # Number of data points
    n = length(x_values)
    # Finding  the interval which contains x
    i = 1
    while i <= n
        if x < x_values[i]
            break
        end
        i += 1
    end
    # If x is outside the data range, return 0.0
    if i == 1 || i == n + 1
        println("x is outside the data range and hence the function will show 0 as interpolated value")
        return 0.0
    end
    # Linear interpolation
    y = y_values[i - 1] + (y_values[i] - y_values[i - 1]) * (x - x_values[i - 1]) / (x_values[i] - x_values[i - 1])
    return y
end
@printf(" Intrapolated value =%16.12f in linear_interpolation method at x  = %4.2f  \n", linear_interpolation(x_datas,y_datas,x1),x1)
@printf(" Intrapolated  value =%16.12f in linear_interpolation method at x  = %4.2f   \n", linear_interpolation(x_datas,y_datas,x2),x2)

File: computational_physics/Homework1/hw_1_1.cpp
#include <iostream>
#include <vector>
#include <cmath>

// Function for nth order polynomial interpolation
double nth_order_interpolation(double x, std::vector<double> x_data, std::vector<double> y_data) {
    int n = x_data.size() - 1;
    double interpolated_value = 0.0;
    
    for (int i = 0; i <= n; ++i) {
        double term = y_data[i];
        
        for (int j = 0; j <= n; ++j) {
            if (i != j) {
                term *= (x - x_data[j]) / (x_data[i] - x_data[j]);
            }
        }
        
        interpolated_value += term;
    }
    
    return interpolated_value;
}

// Function to calculate the error for nth order polynomial interpolation
double error_for_nth_polynomial(double x, std::vector<double> x_data, std::vector<double> y_data) {
    int n = x_data.size() - 1;
    double error = 0.0;
    double term = nth_order_interpolation(x, x_data, y_data) / std::tgamma(n + 2);
    
    for (int i = 0; i <= n; ++i) {
        term *= (x - x_data[i]);
    }
    
    error += term;
    return error;
}



// Function for Neville's Lagrange interpolation
std::pair<double, std::vector<std::vector<double>>> neville_lagrange_interpolation(double x, std::vector<double> x_data, std::vector<double> y_data) {
    int n = x_data.size() - 1;
    std::vector<std::vector<double>> function_value(n + 1, std::vector<double>(n + 1, 0.0));

    // Initialize the first column of the function_value matrix with y_data
    for (int i = 0; i <= n; ++i) {
        function_value[i][0] = y_data[i];
    }

    // Neville's interpolation
    for (int i = 1; i <= n; ++i) {
        for (int j = i; j <= n; ++j) {
            double numerator1 = (x - x_data[j - i]) * function_value[j][i - 1];
            double numerator2 = (x - x_data[j]) * function_value[j - 1][i - 1];
            double denominator = x_data[j] - x_data[j - i];
            function_value[j][i] = (numerator1 - numerator2) / denominator;
        }
    }

    double approximated_value = function_value[n][n];
    
    return std::make_pair(approximated_value, function_value);
}

double error_neville_lagrange_interpolation(double x, std::vector<double> x_data, std::vector<double> y_data) {
    auto result = neville_lagrange_interpolation(x, x_data, y_data);
    int n = x_data.size() - 1;

    double numerical_uncertainty = 0.5 * (std::abs(result.second[n][n] - result.second[n - 1][n - 1]) +
                                          std::abs(result.second[n][n] - result.second[n][n - 1]));

    return numerical_uncertainty;
}
// Function for linear interpolation
double linear_interpolation(std::vector<double> x_values, std::vector<double> y_values, double x) {
    // Number of data points
    int n = x_values.size();
    
    // Finding the interval which contains x
    int i = 0;
    while (i < n) {
        if (x < x_values[i]) {
            break;
        }
        i++;
    }
    
    // If x is outside the data range, return 0.0
    if (i == 0 || i == n) {
        std::cout << "x is outside the data range and hence the function will show 0 as interpolated value" << std::endl;
        return 0.0;
    }
    
    // Linear interpolation
    double y = y_values[i - 1] + (y_values[i] - y_values[i - 1]) * (x - x_values[i - 1]) / (x_values[i] - x_values[i - 1]);
    return y;
}

int main() {
    std::vector<double> x_datas = {0.2, 0.3, 0.4, 0.5, 0.6};
    std::vector<double> y_datas = {0.0015681, 0.0077382, 0.023579, 0.054849, 0.10696};
    double x1 = 0.16;
    double x2 = 0.46;
    double result1 = nth_order_interpolation(x1, x_datas, y_datas);
    double error1 = error_for_nth_polynomial(x1, x_datas, y_datas);
    double result2 = nth_order_interpolation(x2, x_datas, y_datas);
    double error2 = error_for_nth_polynomial(x2, x_datas, y_datas);

    std::cout << "Polynomial value at x = " << x1 << ": " << result1 << " Error: " << error1 << std::endl;
    std::cout << "Polynomial value at x = " << x2 << ": " << result2 << " Error: " << error2 << std::endl;

    auto result1_ = neville_lagrange_interpolation(x1, x_datas, y_datas);
    std::cout << "Approximated value at x = " << x1 << ": " << result1_.first << std::endl;
    std::cout << "Neville iterations table:" << std::endl;
    for (int i = 0; i < result1_.second.size(); ++i) {
        for (int j = 0; j < result1_.second[i].size(); ++j) {
            std::cout << result1_.second[i][j] << "\t";
        }
        std::cout << std::endl;
    }

    auto result2_ = neville_lagrange_interpolation(x2, x_datas, y_datas);
    std::cout << "Approximated value at x = " << x2 << ": " << result2_.first << std::endl;
    std::cout << "Neville iterations table:" << std::endl;
    for (int i = 0; i < result2_.second.size(); ++i) {
        for (int j = 0; j < result2_.second[i].size(); ++j) {
            std::cout << result2_.second[i][j] << "\t";
        }
        std::cout << std::endl;
    }

    double uncertainty1 = error_neville_lagrange_interpolation(x1, x_datas, y_datas);
    double uncertainty2 = error_neville_lagrange_interpolation(x2, x_datas, y_datas);
    std::cout << "Numerical uncertainty at x = " << x1 << ": " << uncertainty1 << std::endl;
    std::cout << "Numerical uncertainty at x = " << x2 << ": " << uncertainty2 << std::endl;
    double interpolated_value1 = linear_interpolation(x_datas, y_datas, x1);
    double interpolated_value2 = linear_interpolation(x_datas, y_datas, x2);

    std::cout << "Interpolated value at x = " << x1 << ": " << interpolated_value1 << std::endl;
    std::cout << "Interpolated value at x = " << x2 << ": " << interpolated_value2 << std::endl;


    return 0;
}
